### PR TITLE
Add new convention rule for CompileStatic

### PR DESCRIPTION
**Pull request by snimavat**
Add new CompileStatic rule which can be used to ensure that classes are annotated with compile static. and if the class does not need to be compile static, it has to be explicitely annotated with @CompileDynamic

### CODE REVIEW

**Comment by snimavat** [comment]
Please review and comment if it needs any changes.

**Comment by jlstrater** [comment]
what about the case where static compilation is added via Gradle?

**Comment by chrismair** [comment]
I am also interested in the answer to the question from @jlstrater re: what if @CompileStatic is configured through Gradle, across the project. I would assume you then not use/enable this rule, right?

File: src/main/groovy/org/codenarc/rule/convention/CompileStaticRule.groovy [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/*
```
Comment (chrismair) [inline src/main/groovy/org/codenarc/rule/convention/CompileStaticRule.groovy]: Please update to 2019

File: src/main/groovy/org/codenarc/rule/convention/CompileStaticRule.groovy [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import org.codenarc.rule.AbstractAstVisitor
+import org.codenarc.rule.AbstractAstVisitorRule
+
+/**
+ * Enforces classes are annotated either with one of the CompileStatic, GrailsCompileStatic or CompileDynamic
+ * annotations
+ *
+ * @Author Sudhir Nimavat
+ */
+@CompileStatic
```
Comment (chrismair) [inline src/main/groovy/org/codenarc/rule/convention/CompileStaticRule.groovy]: Why is @CompileStatic necessary/useful here? It is certainly not conventional in this code base.

File: src/main/groovy/org/codenarc/rule/convention/CompileStaticRule.groovy [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+class CompileStaticRule extends AbstractAstVisitorRule {
+    int priority = 2
```
Comment (chrismair) [inline src/main/groovy/org/codenarc/rule/convention/CompileStaticRule.groovy]: Please add a blank line at the top of each class.

File: src/test/groovy/org/codenarc/rule/convention/CompileStaticRuleTest.groovy [hunk 4]
```diff
@@ -0,0 +1,94 @@
+/*
```
Comment (chrismair) [inline src/test/groovy/org/codenarc/rule/convention/CompileStaticRuleTest.groovy]: Please update to 2019

File: src/test/groovy/org/codenarc/rule/convention/CompileStaticRuleTest.groovy [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        final SOURCE = '''
+            import groovy.transform.CompileDynamic
+            @CompileDynamic
+            class Test {
+                class AnInnerClass { }
+            }
+         '''
+        assertNoViolations(SOURCE)
+    }
+}
```
Comment (chrismair) [inline src/test/groovy/org/codenarc/rule/convention/CompileStaticRuleTest.groovy]: Add test for Enum.
Add test for abstract class.

File: src/test/groovy/org/codenarc/rule/convention/CompileStaticRuleTest.groovy [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import org.codenarc.rule.Rule
+import org.codenarc.rule.Violation
+import org.junit.Test
+
+/**
+ * Tests for CompileStatic rule.
+ *
+ * @author Sudhir Nimavat
+ */
+class CompileStaticRuleTest extends AbstractRuleTestCase {
```
Comment (chrismair) [inline src/test/groovy/org/codenarc/rule/convention/CompileStaticRuleTest.groovy]: Add blank line at top of class

**Comment by basejump** [comment]
I have been working with @snimavat on this and can answer the question of static compilation being added via Gradle.

1. yes, I would assume that if you were doing gradle compile static that then you would not enable this
2. we had problems with getting Idea IDE editor to pick up the gradle setting where it does read the annotation
3. might be angry monkey memory but I recall the global compileStatic being finiky about overriding it with @CompileDynamic and speicifically trying to override it when usign Grails @GrailsCompileStatic
4. we wanted it to be explicit what the intent is on each class so it was immediately obvious if it was @CompileStatic, @CompileDynamic or @GrailsCompileStatic

**Comment by snimavat** [comment]
@chrismair I have made changes, pl review.

**Comment by chrismair** [comment]
Looks good. Thanks for the updates, and thanks for the contribution.

**Pull request merged**